File: android/src/main/java/com/bridge/NativeBridge.kt
package com.bridge

import android.annotation.SuppressLint
import android.os.Build
import android.webkit.JavascriptInterface
import android.webkit.WebSettings
import android.webkit.WebView
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

/** Receives custom events from the web layer (`NativeBridge.send`). */
typealias WebEventListener = (event: String, payload: Any?, webViewId: String, webView: WebView) -> Unit

/**
 * Central, app-wide bridge. Singleton so capture (network/notifications) can feed
 * data to every attached WebView without the host app writing any glue code.
 */
object NativeBridge {

    private val attachedViews = mutableListOf<WebView>()
    private val webViewIds = java.util.WeakHashMap<WebView, String>()
    private val dataStore = ConcurrentHashMap<String, Any?>()
    private var deviceInfoProvider: () -> Map<String, Any?> = { emptyMap() }
    private var webViewSeq = 0

    val apiCalls = ApiCallStore()
    val notifications = NotificationStore()

    private var globalWebEventListener: WebEventListener? = null
    private val webEventListeners = CopyOnWriteArrayList<WebEventListener>()
    private val perWebViewEventListeners = java.util.WeakHashMap<WebView, (String, Any?) -> Unit>()
    private val emitListeners = CopyOnWriteArrayList<(String, Any?) -> Unit>()

    // ---- Host app helpers (optional) ----------------------------------------

    /** Publish any value the web side can read via NativeBridge.getData(key). */
    fun putData(key: String, value: Any?) = setData(key, value, "native")

    /**
     * Store a value in the shared app-wide store. All attached WebViews receive [onData].
     * [sourceWebViewId] identifies the writer ("native", "wv_1", …).
     */
    fun setData(key: String, value: Any?, sourceWebViewId: String?) {
        if (key.isEmpty()) return
        dataStore[key] = value
        val payload = JSONObject()
            .put("key", key)
            .put("value", wrapValue(value))
        if (sourceWebViewId != null) payload.put("source", sourceWebViewId)
        emit("data", payload)
    }

    fun removeData(key: String, sourceWebViewId: String?) {
        if (key.isEmpty()) return
        dataStore.remove(key)
        val payload = JSONObject()
            .put("key", key)
            .put("value", JSONObject.NULL)
            .put("removed", true)
        if (sourceWebViewId != null) payload.put("source", sourceWebViewId)
        emit("data", payload)
    }

    internal fun webViewId(webView: WebView): String =
        webViewIds.getOrPut(webView) { "wv_${++webViewSeq}" }

    fun setDeviceInfoProvider(provider: () -> Map<String, Any?>) {
        deviceInfoProvider = provider
    }

    internal fun getData(key: String): Any? = dataStore[key]

    internal fun getAllData(): Map<String, Any?> = dataStore.toMap()

    internal fun getDeviceInfo(): Map<String, Any?> = deviceInfoProvider()

    // ---- WebView wiring ------------------------------------------------------

    /**
     * Attach to any existing WebView. The drop-in [BridgeWebView] calls this for
     * you. Safe to call multiple times.
     */
    @SuppressLint("SetJavaScriptEnabled", "AddJavascriptInterface")
    fun attach(webView: WebView) {
        webView.settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            mediaPlaybackRequiresUserGesture = false
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                mixedContentMode = WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE
            }
        }
        webView.addJavascriptInterface(JsInterface(webView), "AndroidBridge")
        if (!attachedViews.contains(webView)) attachedViews.add(webView)
        webViewId(webView)
        inject(webView)
        // Push current app state immediately so web has snapshot on load.
        emit("app.state", AppStateMonitor.currentJson())
    }

    fun detach(webView: WebView) {
        attachedViews.remove(webView)
    }

    /** Inject the web SDK. Call from WebViewClient.onPageStarted for SPA reloads. */
    fun inject(webView: WebView) {
        webView.post {
            webView.evaluateJavascript(BridgeScript.JS, null)
            injectSafeAreaCss(AppStateMonitor.getSafeArea())
        }
    }

    internal fun injectSafeAreaCss(insets: SafeAreaInsets) {
        attachedViews.forEach { wv -> BridgeSafeArea.injectCss(wv, insets) }
    }

    // ---- Outbound events to web ---------------------------------------------

    /** Observe every [emit] / [publishEvent] (e.g. React Native event relay). */
    fun addEmitListener(listener: (event: String, payload: Any?) -> Unit) {
        emitListeners.add(listener)
    }

    fun removeEmitListener(listener: (event: String, payload: Any?) -> Unit) {
        emitListeners.remove(listener)
    }

    fun emit(event: String, payload: Any?) {
        val js = "window.NativeBridge && window.NativeBridge.__emit(" +
            "${JSONObject.quote(event)}, ${wrap(payload)})"
        attachedViews.forEach { wv -> wv.post { wv.evaluateJavascript(js, null) } }
        emitListeners.forEach { listener ->
            try {
                listener(event, payload)
            } catch (_: Exception) {
            }
        }
    }

    internal fun recordApiCall(call: ApiCall) {
        apiCalls.add(call)
        emit("api.call", call.toJson())
    }

    fun recordNotification(record: NotificationRecord) {
        notifications.add(record)
        emit("notification", record.toJson())
    }

    /** Push any custom realtime event to all attached WebViews. */
    fun publishEvent(event: String, payload: Any?) {
        emit(event, payload)
    }

    /** Handle every `NativeBridge.send(event, payload)` from web (app-wide). */
    fun setWebEventListener(listener: WebEventListener?) {
        globalWebEventListener = listener
    }

    fun addWebEventListener(listener: WebEventListener) {
        webEventListeners.add(listener)
    }

    fun removeWebEventListener(listener: WebEventListener) {
        webEventListeners.remove(listener)
    }

    internal fun setWebViewEventListener(
        webView: WebView,
        listener: ((event: String, payload: Any?) -> Unit)?,
    ) {
        if (listener == null) perWebViewEventListeners.remove(webView)
        else perWebViewEventListeners[webView] = listener
    }

    private fun deliverWebEvent(
        event: String,
        payload: Any?,
        webViewId: String,
        webView: WebView,
    ) {
        globalWebEventListener?.invoke(event, payload, webViewId, webView)
        webEventListeners.forEach { listener ->
            try {
                listener(event, payload, webViewId, webView)
            } catch (_: Exception) {
            }
        }
        perWebViewEventListeners[webView]?.invoke(event, payload)
    }

    // ---- Inbound requests from web ------------------------------------------

    private class JsInterface(private val webView: WebView) {
        @JavascriptInterface
        fun postMessage(message: String) {
            val msg = try { JSONObject(message) } catch (e: Exception) { return }
            when (msg.optString("type")) {
                "event" -> {
                    val event = msg.optString("event")
                    if (event.isEmpty()) return
                    val payload = msg.opt("payload")
                    deliverWebEvent(event, payload, webViewId(webView), webView)
                    return
                }
                "request" -> Unit
                else -> return
            }
            val id = msg.optString("id")
            val method = msg.optString("method")
            val params = msg.optJSONObject("params") ?: JSONObject()

            when (method) {
                "bridge.requestPermission" -> {
                    val perm = params.optString("permission")
                    BridgePermissions.request(webView.context.applicationContext, perm) { result ->
                        webView.post { resolve(id, result.toJson()) }
                    }
                    return
                }
                "bridge.canGoBackInWebView" -> {
                    resolve(id, BridgeBackPress.canGoBack(webView))
                    return
                }
                "bridge.goBackInWebView" -> {
                    resolve(id, BridgeBackPress.goBack(webView))
                    return
                }
                "bridge.getSafeAreaInsets" -> {
                    resolve(id, AppStateMonitor.getSafeArea().toJson())
                    return
                }
                "bridge.setApplySafeAreaPadding" -> {
                    val enabled = params.optBoolean("enabled", true)
                    if (webView is BridgeWebView) {
                        webView.setApplySafeAreaPadding(enabled)
                    }
                    resolve(id, JSONObject().put("enabled", enabled))
                    return
                }
                "bridge.setSystemUi" -> {
                    resolve(id, BridgeSystemUi.set(webView.context, params))
                    return
                }
                "bridge.getSystemUi" -> {
                    resolve(id, BridgeSystemUi.get())
                    return
                }
                "bridge.getNotificationSettings" -> {
                    resolve(id, BridgeNotifications.getSettings(webView.context))
                    return
                }
                "bridge.cancelNotification" -> {
                    resolve(id, BridgeNotifications.cancel(webView.context, params))
                    return
                }
                "bridge.cancelAllNotifications" -> {
                    resolve(id, BridgeNotifications.cancelAll(webView.context))
                    return
                }
                "bridge.openNotificationSettings" -> {
                    resolve(id, BridgeNotifications.openSettings(webView.context))
                    return
                }
                "bridge.setWebViewCachePolicy" -> {
                    resolve(id, BridgeWebViewCache.setPolicy(webView, params))
                    return
                }
                "bridge.getWebViewCacheInfo" -> {
                    resolve(id, BridgeWebViewCache.getInfo(webView))
                    return
                }
                "bridge.clearWebViewCache" -> {
                    resolve(id, BridgeWebViewCache.clear(webView, params))
                    return
                }
                "bridge.reloadWebView" -> {
                    resolve(
                        id,
                        BridgeWebViewCache.reload(
                            webView,
                            params.optBoolean("bypassCache", true),
                        ),
                    )
                    return
                }
                "bridge.setCookie" -> {
                    resolve(id, BridgeCookies.setCookie(webView, params))
                    return
                }
                "bridge.setCookies" -> {
                    resolve(id, BridgeCookies.setCookies(webView, params))
                    return
                }
                "bridge.getCookies" -> {
                    resolve(id, BridgeCookies.getCookies(webView, params))
                    return
                }
                "bridge.removeCookie" -> {
                    resolve(id, BridgeCookies.removeCookie(webView, params))
                    return
                }
                "bridge.clearCookies" -> {
                    resolve(id, BridgeCookies.clearCookies(webView, params))
                    return
                }
            }

            if (BridgeDispatcher.dispatchAsync(
                    webView.context.applicationContext,
                    method,
                    params,
                ) { result, error ->
                    webView.post {
                        if (error != null) reject(id, error)
                        else resolve(id, result)
                    }
                }
            ) {
                return
            }

            try {
                resolve(
                    id,
                    BridgeDispatcher.dispatch(
                        webView.context.applicationContext,
                        method,
                        params,
                        webViewId(webView),
                    ),
                )
            } catch (e: Exception) {
                reject(id, e.message ?: "error")
            }
        }

        private fun resolve(id: String, result: Any?) =
            eval("window.NativeBridge.__resolve('$id', ${wrap(result)})")

        private fun reject(id: String, error: String) =
            eval("window.NativeBridge.__reject('$id', ${JSONObject.quote(error)})")

        private fun eval(js: String) = webView.post { webView.evaluateJavascript(js, null) }
    }

    // ---- JSON helpers --------------------------------------------------------

    private fun wrap(v: Any?): String = when (v) {
        null -> "null"
        is JSONObject, is JSONArray -> v.toString()
        is Number, is Boolean -> v.toString()
        is String -> JSONObject.quote(v)
        else -> JSONObject.wrap(v)?.toString() ?: "null"
    }

    private fun wrapValue(v: Any?): Any? = v ?: JSONObject.NULL
}

File: ios/Sources/NativeWebViewBridge/NativeBridge.swift
import Foundation
import WebKit

/// Central, app-wide bridge singleton. Owns capture stores and fans data out to
/// every attached WKWebView, so the host app writes essentially no glue code.
public final class NativeBridge: NSObject {

    public static let shared = NativeBridge()

    private let apiCalls = RingBuffer<ApiCall>(max: 200)
    private let notifications = RingBuffer<NotificationRecord>(max: 100)
    private var dataStore: [String: Any] = [:]
    private let dataLock = NSLock()
    private var attached = NSHashTable<WKWebView>.weakObjects()
    private var webViewIds = NSMapTable<WKWebView, NSString>.weakToStrongObjects()
    private var webViewSeq = 0
    private var globalWebEventHandler: ((String, Any?, String, WKWebView) -> Void)?
    private var perWebViewEventHandlers = NSMapTable<WKWebView, AnyObject>.weakToStrongObjects()
    private var emitListeners: [(String, Any?) -> Void] = []
    private let emitLock = NSLock()

    public typealias WebEventHandler = (String, Any?, String, WKWebView) -> Void
    public typealias EmitListener = (String, Any?) -> Void

    private override init() { super.init() }

    /// Call once in application(_:didFinishLaunchingWithOptions:).
    /// Enables automatic network + notification capture.
    public static func start() {
        BridgeURLProtocol.enable()
        NotificationCapture.enable()
        AppStateMonitor.enable()
    }

    // MARK: - Host app helpers

    /// Publish a value the web side can read via NativeBridge.getData(key).
    public func putData(_ key: String, _ value: Any) {
        setData(key, value, source: "native")
    }

    /// Shared store — all attached WebViews receive `onData` with optional `source` id.
    func setData(_ key: String, _ value: Any?, source: String?) {
        guard !key.isEmpty else { return }
        dataLock.lock(); dataStore[key] = value ?? NSNull(); dataLock.unlock()
        var payload: [String: Any] = ["key": key, "value": value ?? NSNull()]
        if let source = source { payload["source"] = source }
        emit("data", payload)
    }

    func removeData(_ key: String, source: String?) {
        guard !key.isEmpty else { return }
        dataLock.lock(); dataStore.removeValue(forKey: key); dataLock.unlock()
        var payload: [String: Any] = ["key": key, "value": NSNull(), "removed": true]
        if let source = source { payload["source"] = source }
        emit("data", payload)
    }

    func webViewId(for webView: WKWebView) -> String {
        if let existing = webViewIds.object(forKey: webView) as String? { return existing }
        webViewSeq += 1
        let id = "wv_\(webViewSeq)"
        webViewIds.setObject(id as NSString, forKey: webView)
        return id
    }

    // MARK: - WebView wiring

    /// Attach to any WKWebView. The drop-in `BridgeWebView` does this for you.
    public func attach(_ webView: WKWebView) {
        let controller = webView.configuration.userContentController
        controller.removeScriptMessageHandler(forName: "iosBridge")
        controller.add(self, name: "iosBridge")

        // Inject the SDK at document start for every navigation — no web changes.
        let alreadyInjected = controller.userScripts.contains {
            $0.source.contains("__installed")
        }
        if !alreadyInjected {
            let script = WKUserScript(
                source: BridgeScript.js,
                injectionTime: .atDocumentStart,
                forMainFrameOnly: false
            )
            controller.addUserScript(script)
        }
        attached.add(webView)
        _ = webViewId(for: webView)
        emit("app.state", AppStateMonitor.currentDictionary())
        injectSafeAreaCss(AppStateMonitor.safeAreaInsetsValue())
    }

    func injectSafeAreaCss(_ insets: SafeAreaInsets) {
        for webView in attached.allObjects {
            BridgeSafeArea.injectCss(into: webView, insets: insets)
        }
    }

    /// Push any custom realtime event to all attached WebViews (e.g. chat typing, call state).
    public func publishEvent(_ event: String, _ payload: Any) {
        emit(event, payload)
    }

    /// Handle every `NativeBridge.send(event, payload)` from web (app-wide).
    public func setWebEventHandler(_ handler: WebEventHandler?) {
        globalWebEventHandler = handler
    }

    func setWebViewEventHandler(
        _ webView: WKWebView,
        handler: ((String, Any?) -> Void)?
    ) {
        if let handler = handler {
            perWebViewEventHandlers.setObject(handler as AnyObject, forKey: webView)
        } else {
            perWebViewEventHandlers.removeObject(forKey: webView)
        }
    }

    private func deliverWebEvent(
        event: String,
        payload: Any?,
        webViewId: String,
        webView: WKWebView?
    ) {
        guard let webView = webView else { return }
        globalWebEventHandler?(event, payload, webViewId, webView)
        if let perView = perWebViewEventHandlers.object(forKey: webView) as? ((String, Any?) -> Void) {
            perView(event, payload)
        }
    }

    // MARK: - Outbound events

      public func addEmitListener(_ listener: @escaping EmitListener) {
        emitLock.lock()
        emitListeners.append(listener)
        emitLock.unlock()
    }

    public func removeEmitListener(_ listener: @escaping EmitListener) {
        emitLock.lock()
        emitListeners.removeAll { $0 as AnyObject === listener as AnyObject }
        emitLock.unlock()
    }

    public func emit(_ event: String, _ payload: Any) {
        let js = "window.NativeBridge && window.NativeBridge.__emit(\(jsonString(event)), \(jsonString(payload)))"
        runJS(js)
        emitLock.lock()
        let listeners = emitListeners
        emitLock.unlock()
        listeners.forEach { $0(event, payload) }
    }

    func recordApiCall(_ call: ApiCall) {
        apiCalls.add(call)
        emit("api.call", call.toDictionary())
    }

    public func recordNotification(_ record: NotificationRecord) {
        notifications.add(record)
        emit("notification", record.toDictionary())
    }

    // MARK: - Inbound requests (WKScriptMessageHandler)

    private func dispatch(method: String, params: [String: Any], sourceWebViewId: String?) throws -> Any? {
        switch method {
        case "bridge.getApiCalls":
            return filterApiCalls(params).map { $0.toDictionary() }
        case "bridge.getNotifications":
            return notifications.all().map { $0.toDictionary() }
        case "bridge.getData":
            let key = params["key"] as? String ?? ""
            dataLock.lock(); defer { dataLock.unlock() }
            return dataStore[key]
        case "bridge.getAllData":
            dataLock.lock(); defer { dataLock.unlock() }
            return dataStore
        case "bridge.setData":
            let key = params["key"] as? String ?? ""
            setData(key, params["value"], source: sourceWebViewId)
            return ["key": key, "stored": true]
        case "bridge.removeData":
            let key = params["key"] as? String ?? ""
            removeData(key, source: sourceWebViewId)
            return ["key": key, "removed": true]
        case "bridge.getWebViewId":
            return ["id": sourceWebViewId ?? "unknown"]
        case "bridge.getDeviceInfo":
            return DeviceInfo.current()
        case "bridge.getAppState":
            return AppStateMonitor.currentDictionary()
        case "bridge.openSettings":
            return ["opened": BridgePermissions.openAppSettings()]
        case "bridge.getUpiApps":
            return BridgePaymentApps.getUpiApps()
        case "bridge.getPaymentApps":
            return BridgePaymentApps.getPaymentApps()
        case "bridge.canOpenUrl":
            return BridgeIntents.canOpenUrl(params["url"] as? String ?? "")
        case "bridge.openUrl":
            return BridgeIntents.openUrl(params["url"] as? String ?? "")
        case "bridge.launchIntent":
            return BridgeIntents.launchIntent(params)
        case "bridge.queryIntents":
            return BridgeIntents.queryIntents(params)
        case "bridge.openUpiPayment":
            return BridgePaymentApps.openUpiPayment(params)
        case "bridge.buildUpiUri":
            return ["uri": BridgePaymentApps.buildUpiUri(params)]
        case "bridge.getCapabilities":
            return BridgeDevice.getCapabilities()
        case "bridge.getClipboard":
            return BridgeDevice.getClipboard()
        case "bridge.setClipboard":
            return BridgeDevice.setClipboard(params["text"] as? String ?? "")
        case "bridge.vibrate":
            return BridgeDevice.vibrate(params)
        case "bridge.dial":
            return BridgeDevice.dial(params["phone"] as? String ?? "")
        case "bridge.sendSms":
            return BridgeDevice.sendSms(
                phone: params["phone"] as? String ?? "",
                body: params["body"] as? String ?? ""
            )
        case "bridge.openMaps":
            return BridgeDevice.openMaps(params)
        case "bridge.setTorch":
            return BridgeDevice.setTorch(params["enabled"] as? Bool ?? true)
        case "bridge.getDeviceResourceStatus":
            return BridgeDevice.getDeviceResourceStatus()
        case "bridge.releaseDeviceResources":
            return BridgeDevice.releaseDeviceResources(params)
        case "bridge.cancelDeviceOperation":
            return BridgeDevice.cancelDeviceOperation(params)
        case "bridge.getSafeAreaInsets":
            return AppStateMonitor.safeAreaDictionary()
        case "bridge.setSystemUi":
            return BridgeSystemUi.set(params)
        case "bridge.getSystemUi":
            return BridgeSystemUi.get()
        case "bridge.cancelNotification":
            let id = params["id"] as? String ?? ""
            return BridgeNotifications.cancel(identifier: id)
        case "bridge.cancelAllNotifications":
            return BridgeNotifications.cancelAll()
        case "bridge.openNotificationSettings":
            return BridgeNotifications.openSettings()
        case "bridge.getWebViewCacheInfo":
            return BridgeWebViewCache.getInfo(webView: nil)
        case "bridge.setWebViewCachePolicy":
            return BridgeWebViewCache.setPolicy(webView: nil, params: params)
        default:
            throw BridgeError.unknownMethod(method)
        }
    }

    /// Async methods — resolve via callback, not synchronous return.
    func dispatchAsync(method: String, id: String, params: [String: Any]) -> Bool {
        if dispatchDeviceAsync(method: method, id: id, params: params) { return true }
        switch method {
        case "bridge.requestPermission":
            let name = params["permission"] as? String ?? ""
            BridgePermissions.request(name) { result in
                self.runJS("window.NativeBridge.__resolve('\(id)', \(self.jsonString(result.toDictionary())))")
            }
            return true
        case "bridge.getPermissionStatus":
            let name = params["permission"] as? String ?? ""
            BridgePermissions.getStatus(name) { result in
                self.runJS("window.NativeBridge.__resolve('\(id)', \(self.jsonString(result.toDictionary())))")
            }
            return true
        case "bridge.getPermissions":
            BridgePermissions.getAllStatuses { list in
                self.runJS("window.NativeBridge.__resolve('\(id)', \(self.jsonString(list)))")
            }
            return true
        case "bridge.getNotificationSettings":
            BridgeNotifications.getSettings { result in
                self.runJS("window.NativeBridge.__resolve('\(id)', \(self.jsonString(result)))")
            }
            return true
        default:
            return false
        }
    }

    /// Route a bridge method from React Native (same API as the web SDK).
    public func dispatchBridge(
        method: String,
        params: [String: Any],
        completion: @escaping (Any?, Error?) -> Void
    ) {
        if dispatchDeviceAsyncBridge(method: method, params: params, completion: completion) {
            return
        }
        if dispatchAsyncBridge(method: method, params: params, completion: completion) {
            return
        }
        do {
            completion(try dispatch(method: method, params: params, sourceWebViewId: "rn"), nil)
        } catch {
            completion(nil, error)
        }
    }

    private func dispatchDeviceAsync(method: String, id: String, params: [String: Any]) -> Bool {
        let resolve: ([String: Any]?) -> Void = { result in
            self.runJS("window.NativeBridge.__resolve('\(id)', \(self.jsonString(result)))")
        }
        let reject: (Error) -> Void = { error in
            self.runJS("window.NativeBridge.__reject('\(id)', \(self.jsonString(error.localizedDescription)))")
        }
        switch method {
        case "bridge.getCurrentLocation":
            BridgeDevice.getCurrentLocation(params) { result, error in
                if let error = error { reject(error) } else { resolve(result) }
            }
            return true
        case "bridge.takePhoto":
            BridgeDevice.takePhoto(params) { result, error in
                if let error = error { reject(error) } else { resolve(result) }
            }
            return true
        case "bridge.pickImage":
            BridgeDevice.pickImage(params) { result, error in
                if let error = error { reject(error) } else { resolve(result) }
            }
            return true
        case "bridge.pickImages":
            BridgeDevice.pickImages(params) { result, error in
                if let error = error { reject(error) } else { resolve(result) }
            }
            return true
        case "bridge.pickContact":
            BridgeDevice.pickContact { result, error in
                if let error = error { reject(error) } else { resolve(result) }
            }
            return true
        case "bridge.share":
            BridgeDevice.share(params) { result, error in
                if let error = error { reject(error) } else { resolve(result) }
            }
            return true
        default:
            return false
        }
    }

    private func dispatchDeviceAsyncBridge(
        method: String,
        params: [String: Any],
        completion: @escaping (Any?, Error?) -> Void
    ) -> Bool {
        switch method {
        case "bridge.getCurrentLocation":
            BridgeDevice.getCurrentLocation(params) { result, error in completion(result, error) }
            return true
        case "bridge.takePhoto":
            BridgeDevice.takePhoto(params) { result, error in completion(result, error) }
            return true
        case "bridge.pickImage":
            BridgeDevice.pickImage(params) { result, error in completion(result, error) }
            return true
        case "bridge.pickImages":
            BridgeDevice.pickImages(params) { result, error in completion(result, error) }
            return true
        case "bridge.pickContact":
            BridgeDevice.pickContact { result, error in completion(result, error) }
            return true
        case "bridge.share":
            BridgeDevice.share(params) { result, error in completion(result, error) }
            return true
        default:
            return false
        }
    }

    private func dispatchAsyncBridge(
        method: String,
        params: [String: Any],
        completion: @escaping (Any?, Error?) -> Void
    ) -> Bool {
        switch method {
        case "bridge.requestPermission":
            let name = params["permission"] as? String ?? ""
            BridgePermissions.request(name) { result in
                completion(result.toDictionary(), nil)
            }
            return true
        case "bridge.getPermissionStatus":
            let name = params["permission"] as? String ?? ""
            BridgePermissions.getStatus(name) { result in
                completion(result.toDictionary(), nil)
            }
            return true
        case "bridge.getPermissions":
            BridgePermissions.getAllStatuses { list in
                completion(list, nil)
            }
            return true
        default:
            return false
        }
    }

    private func filterApiCalls(_ params: [String: Any]) -> [ApiCall] {
        var calls = apiCalls.all()
        if let urlContains = params["urlContains"] as? String, !urlContains.isEmpty {
            calls = calls.filter { $0.url.range(of: urlContains, options: .caseInsensitive) != nil }
        }
        if let method = params["method"] as? String, !method.isEmpty {
            calls = calls.filter { $0.method.caseInsensitiveCompare(method) == .orderedSame }
        }
        if let limit = params["limit"] as? Int, calls.count > limit {
            calls = Array(calls.suffix(limit))
        }
        return calls
    }

    // MARK: - JSON helpers

    private func jsonString(_ value: Any?) -> String {
        guard let value = value else { return "null" }
        guard let data = try? JSONSerialization.data(withJSONObject: [value]),
              let s = String(data: data, encoding: .utf8) else { return "null" }
        return String(s.dropFirst().dropLast()) // unwrap the [ ] so scalars work too
    }

    private func runJS(_ js: String) {
        DispatchQueue.main.async {
            for webView in self.attached.allObjects {
                webView.evaluateJavaScript(js, completionHandler: nil)
            }
        }
    }

    enum BridgeError: Error, LocalizedError {
        case unknownMethod(String)
        var errorDescription: String? {
            switch self {
            case .unknownMethod(let m): return "Unknown method: \(m)"
            }
        }
    }
}

extension NativeBridge: WKScriptMessageHandler {
    public func userContentController(
        _ controller: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard let body = message.body as? String,
              let data = body.data(using: .utf8),
              let msg = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return }

        let sourceId: String? = {
            guard let wv = message.webView else { return nil }
            return webViewId(for: wv)
        }()

        if msg["type"] as? String == "event" {
            let event = msg["event"] as? String ?? ""
            guard !event.isEmpty, let wv = message.webView else { return }
            deliverWebEvent(
                event: event,
                payload: msg["payload"],
                webViewId: sourceId ?? "unknown",
                webView: wv
            )
            return
        }

        guard msg["type"] as? String == "request",
              let id = msg["id"] as? String,
              let method = msg["method"] as? String
        else { return }

        let params = msg["params"] as? [String: Any] ?? [:]

        if dispatchAsync(method: method, id: id, params: params) { return }

        if let wv = message.webView {
            switch method {
            case "bridge.canGoBackInWebView":
                runJS("window.NativeBridge.__resolve('\(id)', \(jsonString(BridgeBackPress.canGoBack(wv)))")
                return
            case "bridge.goBackInWebView":
                runJS("window.NativeBridge.__resolve('\(id)', \(jsonString(BridgeBackPress.goBack(wv)))")
                return
            case "bridge.setApplySafeAreaPadding":
                let enabled = params["enabled"] as? Bool ?? true
                if let bridge = wv as? BridgeWebView {
                    bridge.setApplySafeAreaPadding(enabled)
                }
                runJS("window.NativeBridge.__resolve('\(id)', \(jsonString(["enabled": enabled])))")
                return
            case "bridge.setWebViewCachePolicy":
                runJS("window.NativeBridge.__resolve('\(id)', \(jsonString(BridgeWebViewCache.setPolicy(webView: wv, params: params)))")
                return
            case "bridge.getWebViewCacheInfo":
                runJS("window.NativeBridge.__resolve('\(id)', \(jsonString(BridgeWebViewCache.getInfo(webView: wv)))")
                return
            case "bridge.clearWebViewCache":
                BridgeWebViewCache.clear(webView: wv, params: params) { result in
                    self.runJS("window.NativeBridge.__resolve('\(id)', \(self.jsonString(result)))")
                }
                return
            case "bridge.reloadWebView":
                let bypass = params["bypassCache"] as? Bool ?? true
                runJS("window.NativeBridge.__resolve('\(id)', \(jsonString(BridgeWebViewCache.reload(webView: wv, bypassCache: bypass)))")
                return
            case "bridge.setCookie":
                do {
                    let result = try BridgeCookies.setCookie(webView: wv, params: params)
                    runJS("window.NativeBridge.__resolve('\(id)', \(jsonString(result)))")
                } catch {
                    runJS("window.NativeBridge.__reject('\(id)', \(jsonString(error.localizedDescription)))")
                }
                return
            case "bridge.setCookies":
                do {
                    let result = try BridgeCookies.setCookies(webView: wv, params: params)
                    runJS("window.NativeBridge.__resolve('\(id)', \(jsonString(result)))")
                } catch {
                    runJS("window.NativeBridge.__reject('\(id)', \(jsonString(error.localizedDescription)))")
                }
                return
            case "bridge.getCookies":
                BridgeCookies.getCookies(webView: wv, params: params) { result in
                    self.runJS("window.NativeBridge.__resolve('\(id)', \(self.jsonString(result)))")
                }
                return
            case "bridge.removeCookie":
                BridgeCookies.removeCookie(webView: wv, params: params) { result in
                    self.runJS("window.NativeBridge.__resolve('\(id)', \(self.jsonString(result)))")
                }
                return
            case "bridge.clearCookies":
                BridgeCookies.clearCookies(webView: wv, params: params) { result in
                    self.runJS("window.NativeBridge.__resolve('\(id)', \(self.jsonString(result)))")
                }
                return
            default:
                break
            }
        }

        do {
            let result = try dispatch(method: method, params: params, sourceWebViewId: sourceId)
            runJS("window.NativeBridge.__resolve('\(id)', \(jsonString(result)))")
        } catch {
            runJS("window.NativeBridge.__reject('\(id)', \(jsonString(error.localizedDescription)))")
        }
    }
}

enum DeviceInfo {
    static func current() -> [String: Any] {
        #if canImport(UIKit)
        return [
            "platform": "ios",
            "model": deviceModel(),
            "osVersion": osVersion(),
            "appBundleId": Bundle.main.bundleIdentifier ?? "",
        ]
        #else
        return ["platform": "ios"]
        #endif
    }
    private static func deviceModel() -> String {
        var sysinfo = utsname(); uname(&sysinfo)
        return withUnsafePointer(to: &sysinfo.machine) {
            $0.withMemoryRebound(to: CChar.self, capacity: 1) { String(validatingUTF8: $0) ?? "" }
        }
    }
    private static func osVersion() -> String {
        let v = ProcessInfo.processInfo.operatingSystemVersion
        return "\(v.majorVersion).\(v.minorVersion).\(v.patchVersion)"
    }
}
